### PR TITLE
Implement inspector for notebooks and consoles.

### PR DESCRIPTION
**Pull request by afshin**
Fixes https://github.com/jupyter/jupyterlab/issues/536

- [x] Move inspector functionality to the top-level.
- [x] Dock inspector panel to application shell.
- [x] Create an inspection handler that is similar to the completion handler.
- [x] Allow consoles to connect to the inspection handler.
- [x] Allow notebooks to connect to the inspection handler.

### CODE REVIEW

**Comment by blink1073** [comment]
Woot!  Reviewing now.

File: src/inspector/handler.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+  /**
+   * Update the inspector based on page outputs from the kernel.
+   *
+   * #### Notes
+   * Payloads are deprecated and there are no official interfaces for them in
+   * the kernel type definitions.
+   * See [Payloads (DEPRECATED)](http://jupyter-client.readthedocs.io/en/latest/messaging.html#payloads-deprecated).
+   */
+  execute(content: KernelMessage.IExecuteOkReply): void {
```
Comment (blink1073) [inline src/inspector/handler.ts]: This name is confusing, perhaps `handleExecuteReply`?

File: src/inspector/handler.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  const disposedSignal = new Signal<InspectionHandler, void>();
+
+  /**
+   * A signal emitted when an inspector value is generated.
+   */
+  export
+  const inspectedSignal = new Signal<InspectionHandler, Inspector.IInspectorUpdate>();
+}
+
+// editor.textChanged.connect(this.onTextChange, this);
```
Comment (blink1073) [inline src/inspector/handler.ts]: Errant comment

File: src/inspector/handler.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+import {
+  ISignal, Signal, clearSignalData
+} from 'phosphor-signaling';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import {
```
Comment (blink1073) [inline src/inspector/handler.ts]: Should be imported last.

File: src/console/plugin.ts [hunk 4]
```diff
@@ -10,14 +10,14 @@ import {
...
 
 import {
-  RenderMime
-} from '../rendermime';
-
-import {
   selectKernel
 } from '../docregistry';
 
 import {
```
Comment (blink1073) [inline src/console/plugin.ts]: Tidy up import order in this file?

File: src/inspector/inspector.ts [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+     */
+    remember?: boolean;
+
+    /**
+     * The type of the inspector.
+     */
+    type: string;
+
+    /**
+     * The optional console inspector widget instance.
```
Comment (blink1073) [inline src/inspector/inspector.ts]: Remove the word "console" in this file.

File: src/inspector/inspector.ts [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+   * A signal emitted when an inspector widget's orientation is toggled.
+   */
+  get orientationToggled(): ISignal<InspectorItem, void> {
+    return Private.orientationToggledSignal.bind(this);
+  }
+
+  /**
+   * A flag that indicates whether the inspector remembers history.
+   */
+  get remember(): boolean {
```
Comment (blink1073) [inline src/inspector/inspector.ts]: I don't like using verbs for properties, maybe `remembers`?

File: src/notebook/notebook/widget.ts [hunk 7]
```diff
@@ -547,6 +559,26 @@ namespace StaticNotebook {
  */
 export
 class Notebook extends StaticNotebook {
+  constructor(options: StaticNotebook.IOptions) {
```
Comment (blink1073) [inline src/notebook/notebook/widget.ts]: Missing docstring

File: src/notebook/notebook/widget.ts [hunk 8]
```diff
@@ -547,6 +559,26 @@ namespace StaticNotebook {
...
+    super(options);
+    // Set up the inspection handler.
+    this._inspectionHandler = new InspectionHandler(this.rendermime);
+    this.activeCellChanged.connect((s, cell) => {
+      this._inspectionHandler.activeCell = cell;
+    });
+  }
+
+
```
Comment (blink1073) [inline src/notebook/notebook/widget.ts]: Extra line

**Comment by blink1073** [comment]
noice

**Comment by jasongrout** [comment]
For future reference, here is a demo of this PR.
inspector

**Comment by afshin** [comment]
Thanks, @jasongrout

**Pull request merged**